---
title: "MutationalPatterns"
date: "`r Sys.time()`"
output:
  html_document:
    code_folding: hide
resource_files:
  - outputs/sig_plots/Sig
  - outputs/sig_plots/SBS
  - outputs/sig_plots/DBS
  - outputs/sig_plots/ID
---

```{r knitr_opts, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE}
require(MutationalPatterns)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg38"
require(ref_genome, character.only = TRUE)
#require(devtools)
require(dplyr)
require(patchwork)
require(kableExtra)
require(deconstructSigs)
```

## Data Preparation

```{r data_prep}
params <- list(
  vcf = "~/Desktop/research_work/paad/signature/vcf/final_report.vcf",
  nm = "01SkNC1D9.1",
  outdir = "~/Desktop/research_work/paad/signature/output"
)

gr <- MutationalPatterns::read_vcfs_as_granges(
  vcf_files = params$vcf,
  sample_names = params$nm,
  genome = ref_genome,
  group = "auto+sex",
  type = "all"
)
```
```{r define_sig_count_snv, include=FALSE}
sig_count_snv <- function(vcf_gr, ref_genome) {
  gr_snv <- MutationalPatterns::get_mut_type(vcf_list = vcf_gr, type = "snv")
  snv_counts <- MutationalPatterns::mut_matrix(vcf_list = gr_snv, ref_genome = ref_genome)
  list(
    snv_counts = snv_counts,
    gr_snv = gr_snv
  )
}
sig_plot_snv <- function(gr_snv, snv_counts, ref_genome) {
  mut_to <- MutationalPatterns::mut_type_occurrences(
    vcf_list = gr_snv, ref_genome = ref_genome
  )
  
  mut_mat_ext_context <- MutationalPatterns::mut_matrix(
    vcf_list = gr_snv, ref_genome = ref_genome, extension = 2
  )
  
  p_spectrum <- MutationalPatterns::plot_spectrum(
    type_occurrences = mut_to, CT = TRUE,
    condensed = TRUE, error_bars = "none"
  ) +
    ggplot2::theme(legend.position = "top")
  p_96_profile <- MutationalPatterns::plot_96_profile(mut_matrix = snv_counts, condensed = TRUE)
  p_heatmap <- MutationalPatterns::plot_profile_heatmap(mut_matrix = mut_mat_ext_context) +
    ggplot2::theme(legend.position = "none")
  p_river <- MutationalPatterns::plot_river(mut_matrix = mut_mat_ext_context) +
    ggplot2::theme(legend.position = "none")
  
  list(
    p_heatmap = p_heatmap,
    p_river = p_river,
    p_96_profile = p_96_profile,
    p_spectrum = p_spectrum
  )
}
```

## SNVs (SBS)

### Plots

```{r snv_plots, fig.width=12, fig.height=10, out.width="90%"}
p_snvs <- sig_plot_snv(
  gr_snv = snv_counts$gr_snv, snv_counts = decon_data,
  ref_genome = ref_genome
)
p_snvs$p_river + p_snvs$p_heatmap + p_snvs$p_spectrum + p_snvs$p_96_profile + patchwork::plot_layout(ncol = 2)
```

```{r MutationalPatterns_SBS_plot, include=FALSE}
p_snvs <- sig_plot_snv(
  gr_snv = snv_counts$gr_snv, snv_counts = snv_counts$snv_counts,
  ref_genome = ref_genome
)

p_snvs$p_96_profile
library(ggplot2)
ggsave("snv_96_profile_mp_filter.pdf", plot = p_snvs$p_96_profile, width = 9, height = 4)
```

```{r deconstructSigs_SBS_plot, include=FALSE}
snv_counts <- sig_count_snv(vcf_gr = gr, ref_genome = ref_genome)
tumor_ref_df <- as.data.frame(t(snv_counts[["snv_counts"]]))

sample_1 = deconstructSigs::whichSignatures(tumor.ref = tumor_ref_df, 
                           signatures.ref = signatures.nature2013, 
                           sample.id = params$nm, 
                           contexts.needed = TRUE,
                           tri.counts.method = "genome")
# mut_matrix
decon_data<- sample_1[["tumor"]]
decon_data <- t(decon_data)

p_snvs <- sig_plot_snv(
  gr_snv = snv_counts$gr_snv, snv_counts = decon_data,
  ref_genome = ref_genome
)

p_snvs$p_96_profile
library(ggplot2)
ggsave("snv_96_profile_ds_genome_filter.pdf", plot = p_snvs$p_96_profile, width = 9, height = 4)
```


```{r define_sig_contribution, include=FALSE}
sig_contribution <- function(mut_mat, signatures) {
  # Fit mutation matrix to cancer signatures
  fit_res <-
    MutationalPatterns::fit_to_signatures(mut_mat, signatures)$contribution |>
    tibble::as_tibble(rownames = "sig") |>
    dplyr::rename(contr = 2) |>
    dplyr::filter(.data$contr > 0)
  
  if (nrow(fit_res) == 0) {
    fit_res <- tibble::tribble(
      ~sig, ~contr,
      "No Signatures found!", 0
    )
  }
  
  fit_res_contr <- fit_res |>
    dplyr::mutate(
      contr = round(.data$contr, 3),
      RelFreq = round(.data$contr / sum(.data$contr), 2),
      Rank = as.integer(base::rank(-.data$contr))
    ) |>
    dplyr::select("Rank",
                  Signature = "sig",
                  Contribution = "contr", "RelFreq"
    ) |>
    dplyr::arrange(.data$Rank)
  
  fit_res_contr
}
sig_contribution_table <- function(contr, type, outdir = NULL) {
  available_types <- c(
    "Sig" = "v2_2015/Sig",
    "SBS" = "v3.2_2021-march/SBS",
    "DBS" = "v3.2_2021-march/DBS",
    "ID" = "v3.2_2021-march/ID"
  )
  assertthat::assert_that(length(type) == 1, type %in% names(available_types))
  assertthat::assert_that(all(colnames(contr) == c("Rank", "Signature", "Contribution", "RelFreq")))
  
  #sig_dir <- system.file(file.path("extdata/sigs", available_types[type]), package = "sigrap")
#  sig_dir <- '~/Downloads/'
  sig_dir <- file.path(
    available_types[type]
  )
  path_home <- path.expand("~")
  outdir <- file.path("output")
  # if an outdir is specified, copy out the COSMIC plots to be linked to
  if (!is.null(outdir)) {
    img_cp_dir <- file.path(outdir, "sig_plots", type)
    fs::dir_create(img_cp_dir)
    sig_table <-
      readr::read_tsv(file = file.path(sig_dir, "description.tsv"), col_types = "cc") |>
      dplyr::mutate(
        Plot_original = file.path(sig_dir, paste0(.data$signature, ".png")),
        Plot_copy = file.path(img_cp_dir, paste0(.data$signature, ".png")),
        signature = paste0(type, .data$signature)
      ) |>
      dplyr::rename(Signature = .data$signature)
    
    d <- contr |>
      dplyr::left_join(sig_table, by = "Signature") |>
      dplyr::rowwise() |>
      dplyr::mutate(cp = file.copy(from = .data$Plot_original, to = .data$Plot_copy)) |>
      dplyr::mutate(plot_in_md = paste0("<img src='", .data$Plot_copy, "' width='200px'/>")) |>
      dplyr::select(
        "Rank", "Signature", "Contribution", "RelFreq",
        Description = "description", Plot = "plot_in_md"
      )
  } else { # don't copy out any COSMIC plots
    sig_table <-
      readr::read_tsv(file = file.path(sig_dir, "description.tsv"), col_types = "cc") |>
      dplyr::mutate(
        signature = paste0(type, .data$signature)
      ) |>
      dplyr::rename(Signature = .data$signature)
    d <- contr |>
      dplyr::left_join(sig_table, by = "Signature") |>
      dplyr::rename(Description = .data$description)
  }
  return(d)
}
```


### Signature Contributions

```{r snv_contributions}
sigs_snv_2020 <-
  MutationalPatterns::get_known_signatures(
    muttype = "snv",
    incl_poss_artifacts = TRUE
  ) |>
  {
    \(sigs) sig_contribution(mut_mat = snv_counts$snv_counts, signatures = sigs)
  }()

sigs_snv_2020 |>
  sig_contribution_table(type = "SBS", outdir = params$outdir) |>
  knitr::kable(format = "html", escape = FALSE) |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")
```
```{r define_sig_count_dbs, include=FALSE}
sig_count_dbs <- function(vcf_gr) {
  gr = vcf_gr[[1]]
  dbs_i = elementNROWS(gr$REF) == 2 & sapply(gr$ALT, function(x) nchar(x) == 2)
  dbs_predetermined_gr = gr[dbs_i]

  if (length(dbs_predetermined_gr) == 0) {
    warning("No DBS mutations found in the VCF file.")
    return(NULL)
  }

  gr_dbs <- MutationalPatterns::get_dbs_context(vcf_list = dbs_predetermined_gr)

  dbs_counts <- MutationalPatterns::count_dbs_contexts(vcf_list = gr_dbs)

  return(dbs_counts)
}

sig_plot_dbs <- function(dbs_counts) {
  p_dbs_main <- MutationalPatterns::plot_main_dbs_contexts(counts = dbs_counts)
  p_dbs_cont <- MutationalPatterns::plot_dbs_contexts(counts = dbs_counts, condensed = TRUE)
  
  list(
    p_dbs_main = p_dbs_main,
    p_dbs_cont = p_dbs_cont
  )
}

```

## DBS

### Plots

```{r dbs_plots, fig.width=12, fig.height=10, out.width="90%"}
dbs_counts <- sig_count_dbs(vcf_gr = gr)
p_dbs <- sig_plot_dbs(dbs_counts = dbs_counts)
p_dbs$p_dbs_main / p_dbs$p_dbs_cont
```

### Signature Contributions

```{r dbs_contributions}
sigs_dbs <-
  MutationalPatterns::get_known_signatures(muttype = "dbs") |>
  {
    \(sigs) sig_contribution(mut_mat = dbs_counts, signatures = sigs)
  }()
sigs_dbs |>
  sig_contribution_table(type = "DBS", outdir = params$outdir) |>
  knitr::kable(format = "html", escape = FALSE) |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")
```
```{r define_sig_count_indel, include=FALSE}
sig_count_indel <- function(vcf_gr, ref_genome) {
  gr_indel <- MutationalPatterns::get_mut_type(vcf_list = vcf_gr, type = "indel")
  gr_indel <- MutationalPatterns::get_indel_context(vcf_list = gr_indel, ref_genome = ref_genome)
  indel_counts <- MutationalPatterns::count_indel_contexts(vcf_list = gr_indel)
  indel_counts
}

sig_plot_indel <- function(indel_counts) {
  p_indel_main <- MutationalPatterns::plot_main_indel_contexts(counts = indel_counts) +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 30, vjust = 0.9, hjust = 1)) +
    ggplot2::theme(axis.text.x = ggplot2::element_blank())
  p_indel_cont <- MutationalPatterns::plot_indel_contexts(counts = indel_counts, condensed = TRUE) +
    ggplot2::theme(
      axis.text.x = ggplot2::element_text(angle = 30, vjust = 0.9, hjust = 1),
      legend.position = "top"
    )
  
  list(
    p_indel_main = p_indel_main,
    p_indel_cont = p_indel_cont
  )
}
```

## INDELs

### Plots

```{r indel_plots, fig.width=12, fig.height=10, out.width="90%"}
indel_counts <- sig_count_indel(vcf_gr = gr, ref_genome = ref_genome)
p_indels <- sig_plot_indel(indel_counts = indel_counts)
p_indels$p_indel_main / p_indels$p_indel_cont
```

### Signature Contributions

```{r indel_contributions}
sigs_indel <-
  MutationalPatterns::get_known_signatures(muttype = "indel") |>
  {
    \(sigs) sig_contribution(mut_mat = indel_counts, signatures = sigs)
  }()
sigs_indel |>
  sig_contribution_table(type = "ID", outdir = params$outdir) |>
  knitr::kable(format = "html", escape = FALSE) |>
  kableExtra::kable_styling(c("hover", "striped"), font_size = 12) |>
  kableExtra::scroll_box(height = "400px")
```