---
title: "Tumour Mutational Burden (TMB)"
author: "Sehrish Kanwal"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    keep_md: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE) # global setup
```

Tumour mutational burden (TMB) "measures the quantity of mutations found in a tumor" (https://www.focr.org/tmb). Also, it is defined as "a quantitative measure of the total number of mutations per coding area of a tumor genome." (https://www.genengnews.com/gen-exclusives/the-promise-of-tumor-mutational-burden-for-cancer-immunotherapy/77900833).
This type of biomarker is currently under study to evaluate whether it may help predict the likelihood a patient's response to immunotherapy in a range of advanced cancers. 

Tumors that have higher levels of TMB are believed to express more neoantigens – a type of cancer-specific antigen – that may allow for a more robust immune response and therefore a more durable response to immunotherapy.

Required packages

```{r, message=FALSE}
library(vcfR)
library(knitr)
library(kableExtra)
```

```{r}
#vcf <-  read.vcfR("~/Documents/UMCCR/data/vcfs/ensemble-with_chr_prefix.vcf", verbose =  FALSE) 
vcf_file <- "~/Documents/UMCCR/data/vcfs/ensemble-pon-pass.vcf"
bcftools <- "bcftools query -f '%CHROM\t%POS\t%INFO/ANN\n'"
vcf <- system(paste(bcftools, vcf_file), intern = TRUE) %>%
  tibble::tibble(all_cols = .) %>%
  tidyr::separate(col = .data$all_cols,
                  into = c("CHROM", "POS", "ANN"),
                  sep = "\t", convert = TRUE)

#Calculating mutations per megabase
fix <- getFIX(vcf)
vcf_number_rows <- nrow(fix)
mutations_megabase <- round(vcf_number_rows/3200, digits = 2)

#Summarizing annotations for variants in the vcf
ann <- vcfR::extract.info(vcf, "ANN")
region_ann <- sapply(ann, function(x){
  y <- strsplit(x, "\\|")[[1]][2]
})
variant_annotation <- unname(region_ann)

#Creating a nice table output for annotations summary
region_ann_df <- data.frame(table(variant_annotation))
kable(region_ann_df, caption = "Table summarizing all annotations in the vcf and the total number of variants suppporting these annotations")  %>%
  kable_styling(font_size = 12, "striped", "bordered")
  
#Calculating mutations per megabase in coding region
coding_variants = 0
coding_variants <- region_ann %in% c("frameshift_variant", "missense_variant", "missense_variant&splice_region_variant")
coding_variants <- table(coding_variants)
mutations_megabase_coding <- round(as.vector(coding_variants[2])/40, digits = 2) 
#40MB is the estimated size of coding region in human genome - as used by PCGR as well. 
#We can use 36MB if we go with exact calculations, as only 1.2% of the total genome is considered coding. 
#total genome * percent protein coding = 3,000,000,000 * 0.012 = 36,000,000 ~36MB

#Displaying results in a table
region <- c("Wholegenome", "Coding")
total_mutations <- c(vcf_number_rows, as.vector(coding_variants[2]))
mutations_mb <- c(mutations_megabase, mutations_megabase_coding)

result_display <- data.frame(region, total_mutations, mutations_mb) 
#kable(result_display,  caption = "Table summarizing somatic burden result")

kable(result_display, caption = "Table summarizing somatic burden result") %>%
  kable_styling(font_size = 12, "striped", "bordered")
```

* The _total number of mutations_ in the vcf are **`r vcf_number_rows`** and 
* Number of mutations per megabase are **`r mutations_megabase`**.
* The _total number of mutations in the coding region_ are **`r as.vector(coding_variants[2])`**
* Number of mutations per megabase in the coding region are **`r mutations_megabase_coding`**

## Addendum

```{r sessioninfo}
devtools::session_info()
```
