---
title: "Tumour Mutational Burden (TMB)"
author: "Sehrish Kanwal"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    keep_md: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE) # global setup
```

Tumour mutational burden (TMB) "measures the quantity of mutations found in a tumor" (https://www.focr.org/tmb). Also, it is defined as "a quantitative measure of the total number of mutations per coding area of a tumor genome." (https://www.genengnews.com/gen-exclusives/the-promise-of-tumor-mutational-burden-for-cancer-immunotherapy/77900833).
This type of biomarker is currently under study to evaluate whether it may help predict the likelihood a patient's response to immunotherapy in a range of advanced cancers.

As multiple definitions indicate, currently, there is a lack of standardization for TMB calculation and reporting. However, it is imperative to use this information to further inform clinically-meaningful results, which will support informed decision-making for patients. 

Required packages

```{r, message=FALSE}
library(vcfR, warn.conflicts=F, quietly=T)
library(dplyr, warn.conflicts=F, quietly=T)
library(ggplot2, warn.conflicts=F, quietly=T)
library(reticulate, warn.conflicts=F, quietly=T)
```

```{r}
vcf <-  read.vcfR("~/Documents/UMCCR/data/ensemble-pon-pass.vcf") 

#Calculating mutations per megabase
fix <- getFIX(vcf)
vcf_number_rows <- nrow(fix)
mutations_megabase <- ceiling(vcf_number_rows/3200)

#Summarizing annotations for variants in the vcf
ann <- vcfR::extract.info(vcf, "ANN")
region_ann <- sapply(ann, function(x){
  y <- strsplit(x, "\\|")[[1]][2]
})
region_ann <- unname(region_ann)
table(region_ann)

#Calculating mutations per megabase in coding region
coding_variants = 0
coding_variants <- ifelse(region_ann %in% c("frameshift_variant", "missense_variant", "missense_variant&splice_region_variant"),
       TRUE,
       FALSE)
coding_variants <- table(coding_variants)
mutations_megabase_coding <- ceiling(as.vector(coding_variants[2])/3200)
```

* The _total number of mutations_ in the vcf are **`r vcf_number_rows`** and 
* Number of mutations per megabase are **`r mutations_megabase`**.
* The _total number of mutations in the coding region_ are **`r as.vector(coding_variants[2])`**
* Number of mutations per megabase in the coding region are **`r mutations_megabase_coding`**







