---
title: "Varlap exploration"
author: "Sehrish Kanwal"
date: "27/03/2020"
output:
  html_document:
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(vcfR)
library(dplyr)
library(here)
```

## Overview

The basic idea here is to compare bcbio and Dragen specific variants using [varlap](https://github.com/bjpop/varlap).

## Input Data

- Working with P025 data in `/g/data3/gx8/projects/Hofmann_Cromwell/2019-02-01T0241_Cromwell_WGS_2016.249.18.WH.P025/final`.

- And the vcfs being tested against are in `/g/data/gx8/projects/Hofmann_Cromwell/Dragen/P025/var_cmp/isec_filt_pass/`

```
0000.vcf for records private to bcbio,
0001.vcf for records private to dragen,
0002.vcf for shared records
```

## Processing

For varlap, to be able to include a Class column, for each variant, the input variants must be in CSV format.

Relatively straightforward to achieve using following steps - scripted [here](https://github.com/umccr/sk-explore/blob/master/varlap/analysis/vcf2csv.R). The script performs the following steps:

- drop all the VCF lines starting with #

- remove all unneeded columns

- add Class column

- add a new header row

- required columns are:

- chrom,pos,ref,alt

- CSV header names should be as above, lowercase

- any additional columns can be added in any case

```{r , echo=FALSE}
bcbio_df <- read.csv("~/Documents/UMCCR/data/projects/varlap/results/bcbio_fix.csv", header = TRUE)
dragen_df <- read.csv("~/Documents/UMCCR/data/projects/varlap/results/dragen_fix.csv", header = TRUE)
```

### Processed data summary

Varlap considers the genomic locus of each variant in each of the supplied BAM files and records information about the corresponding alignment context at that locus. Varlap outputs a CSV file containing one row per input variant, with columns recording the various computed metrics about that variant.  

```{r, echo=FALSE}
varlap_output <- read.csv("~/Documents/UMCCR/data/projects/varlap/results/variants.bcbioFIX.csv", header = TRUE)
names(varlap_output)
```

**bcbio specific variants**

```{r bcbio, echo=FALSE}
dim(bcbio_df)
head(bcbio_df)
```

**dragen specific variants**

```{r dragen, echo=FALSE}
dim(dragen_df)
head(dragen_df)
```

## Including Plots

### Distribution

The distribution of variants between both bcbio (left) and dragen (right) classes looks fairly similar

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/varlap_comb_fix.tumour.all.avg.NM.class.dist.png")
knitr::include_graphics("../images/varlap_comb_fix.tumour.alt.vaf.class.dist.png")
```

### Histograms

**pos.normalized**  bcbio (left) and dragen (right)

It is an output from varlap that tells a normalised chromosome position for a variant between [0,1]. Basically it takes the genomic position of the variant (on a given chromosome) and divides it by the length of the chromosome.

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/variants.bcbioFIX2.pos.normalised.histogram.png")
knitr::include_graphics("../images/variants.dragenFIX2.pos.normalised.histogram.png")
```

**tumour.alt.vaf**  bcbio (left) and dragen (right)

tumour.alt.vaf is the VAF of the ALT allele at this locus in the tumour sample. VAF is: num(ALT) / total_reads

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/variants.bcbioFIX2.tumour.alt.vaf.histogram.png")
knitr::include_graphics("../images/variants.dragenFIX2.tumour.alt.vaf.histogram.png")
```

**tumour.all.avg.base.qual**  bcbio (left) and dragen (right)

tumour.all.avg.base.qual is the average base quality at this locus for ALL reads overlapping this locus. By ALL reads it means it doesn't matter whether the read contains the REF or ALT allele (or anything else). It just counts ALL reads.

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/variants.bcbioFIX2.tumour.all.avg.base.qual.histogram.png")
knitr::include_graphics("../images/variants.dragenFIX2.tumour.all.avg.base.qual.histogram.png")
```

**tumour.all.avg.map.qual**  bcbio (left) and dragen (right)

tumour.all.avg.map.qual is the average mapping quality of reads that overlap the locus. By ALL reads it means it doesn't matter whether the read contains the REF or ALT allele (or anything else). It just counts ALL reads.

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/variants.bcbioFIX2.tumour.all.avg.map.qual.histogram.png")
knitr::include_graphics("../images/variants.dragenFIX2.tumour.all.avg.map.qual.histogram.png")
```

**tumour.alt.avg.base.qual**  bcbio (left) and dragen (right)

tumour.alt.avg.base.qual is the base quality of reads that overlap the locus, but only considering reads that support the ALT allele, in the tumour sample.

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/variants.bcbioFIX2.tumour.alt.avg.base.qual.histogram.png")
knitr::include_graphics("../images/variants.dragenFIX2.tumour.alt.avg.base.qual.histogram.png")
```

**tumour.alt.avg.map.qual**  bcbio (left) and dragen (right)

tumour.alt.avg.map.qual is the average mapping quality of reads that overlap the locus, but only considering reads that support the ALT allele, in the tumour sample.

```{r, echo=FALSE, fig.show = "hold", out.width = "50%"}
knitr::include_graphics("../images/variants.bcbioFIX2.tumour.alt.avg.map.qual.histogram.png")
knitr::include_graphics("../images/variants.dragenFIX2.tumour.alt.avg.map.qual.histogram.png")
```

### Scatter

As currenly, I have run varlap with tumor only mode, so compared the following variables:

**tumour.ref.avg.NM and tumour.alt.avg.NM** 

This is the average edit distance of reads overlapping the locus for reads supporting the ref/alt allele in the tumour sample

```{r, echo=FALSE}
knitr::include_graphics("../images/varlap_comb_fix.tumour.ref.avg.NM.tumour.alt.avg.NM.scatter.png")
```


## Interesting follow-up pointers

- Generate the output for variants at AF < 0.15 that are either unique to bcbio or to Dragen.
- Check the differance between variants called just by bcbio vs those that are shared (globally or at low AF).

